### PR TITLE
Provide suggestion for missing `>` in a type parameter list

### LINKED ISSUE #94058: Bad diagnostics span for unmatched angle brackets of generic args
Given the following code: [Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2792a5be7755440d66bdb9dbb0f05c47)

```rust
use std::sync::{Arc, Mutex};

pub struct Foo {
    a: Mutex<usize>,
    b: Arc<Mutex<usize>, // an not-ok line
    c: Arc<Mutex<usize>>, // an ok line
}
```

The current output is:

```
error: expected one of `>`, a const expression, lifetime, or type, found `}`
 --> src/lib.rs:7:1
  |
6 |     c: Arc<Mutex<usize>>, // an ok line
  |                          - expected one of `>`, a const expression, lifetime, or type
7 | }
  | ^ unexpected token
```

Ideally the output should look like:

```
error: expected one of `>`, a const expression, lifetime, or type, found `}`
 --> src/lib.rs:6:1
  |
5 |     b: Arc<Mutex<usize>, // an not-ok line
  |                         - expected one of `>`, a const expression, lifetime, or type
6 | }
  | ^ unexpected token
```

### PR DESCRIPTION
**Pull request by estebank**
When encountering an inproperly terminated type parameter list, provide
a suggestion to close it after the last non-constraint type parameter
that was successfully parsed.

Fix #94058.

### CODE REVIEW

**Comment by compiler-errors** [comment]
I am somewhat concerned about the ratio of misleading to useful "add `>`" suggestions here

... though I'm not sure how much additional effort it would take to improve this to e.g. not suggest adding `>` when there are already more `>` tokens downstream, or e.g. not suggest changing `<a; b>` into `<a>; b>` when the user uses an incorrect delimiter in a comma-separated list.

For example:
 ```
    |
help: you might have meant to end the type parameters here
    |
LL |     fn f1<'a>(arg : Box<dyn X<X::Y> = u32>>) {}
    |                                   +
```

```
   |
help: you might have meant to end the type parameters here
   |
LL | fn foo<'a, 'b>(x: &mut Foo<'a>; 'b>) {}
   |                              +
```

**Comment by estebank** [comment]
@JohnCSimon I'm gonna revert the labels back here because I'm waiting on further review 😅

I feel like the fallout is acceptable as is for merging, but can understand if the reviewers disagree.

**Review by jackh726 — APPROVED** [review]
I think it's hard to say anything about the "very wrong syntax" cases. But, I think this is worth it from the new tests, which more likely represent what a user would normally run into.

One nit, but r=me with/without.

**Comment by estebank** [comment]
Added one more check for `,`/`;` typo in type params.

File: compiler/rustc_parse/src/parser/path.rs [hunk 1]
```diff
@@ -272,7 +272,23 @@ impl<'a> Parser<'a> {
...
+                            .iter()
+                            .rev()
+                            .skip_while(|arg| matches!(arg, AngleBracketedArg::Constraint(_)))
+                            .next()
+                        {
+                            err.span_suggestion_verbose(
+                                arg.span().shrink_to_hi(),
+                                "you might have meant to end the type parameters here",
+                                ">".to_string(),
+                                Applicability::MaybeIncorrect,
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_parse/src/parser/path.rs:287]: I think it's the best of the available `Applicability`s we have available today :)

**Comment by estebank** [comment]
Looks like a timeout error?

```
/checkout/src/test/rustdoc-gui/search-tab-selection-if-current-is-empty.goml search-tab-selection-if-current-is-empty... FAILED
Error:  (line 6) TimeoutError: waiting for selector "#titles" failed: timeout 30000ms exceeded: for command `wait-for: "#titles"`

Build completed unsuccessfully in 0:00:46
== clock drift check ==
  local time: Sun Mar 27 15:02:11 UTC 2022
  network time: Sun, 27 Mar 2022 15:02:11 GMT
== end clock drift check ==
F
Error: Process completed with exit code 1.
```

@bors retry

**Pull request merged**